org.eclipse.xtext.xbase.ui/src/org/eclipse/xtext/xbase/ui/editor/XbaseEditorInputRedirector.java

```diff
@@ -0,0 +1,165 @@
+import org.eclipse.jdt.core.ITypeRoot;
+import org.eclipse.jdt.core.JavaCore;
+import org.eclipse.jdt.core.JavaModelException;
+import org.eclipse.ui.IEditorInput;
+import org.eclipse.ui.ide.ResourceUtil;
+import org.eclipse.ui.part.FileEditorInput;
+import org.eclipse.xtext.LanguageInfo;
+import org.eclipse.xtext.resource.FileExtensionProvider;
+import org.eclipse.xtext.ui.editor.utils.EditorUtils;
+import org.eclipse.xtext.ui.generator.trace.IEclipseTrace;
+import org.eclipse.xtext.ui.generator.trace.ILocationInEclipseResource;
+import org.eclipse.xtext.ui.generator.trace.ITraceForStorageProvider;
+
+import com.google.common.base.Optional;
+import com.google.common.base.Preconditions;
+import com.google.common.collect.FluentIterable;
+import com.google.common.collect.Iterables;
+import com.google.inject.Inject;
+
+public class XbaseEditorInputRedirector {
+
+	private final static Logger LOG = Logger.getLogger(XbaseEditorInputRedirector.class);
+
+	@Inject
+	private ITraceForStorageProvider traceInformation;
+
+	@Inject
+	private LanguageInfo languageInfo;
+
+	@Inject
+	private FileExtensionProvider fileExtensionProvider;
+
+	public ITypeRoot getTypeRoot(IEditorInput editorInput) {
+		IJavaElement adapter = editorInput.getAdapter(IJavaElement.class);
+		return adapter instanceof ITypeRoot ? (ITypeRoot) adapter : null;
+	}
+
+	/**
+	 * @return the original source for an editor input that points to an Xtext resource copied to the output folder, the given input
+	 *         otherwise
+	 */
+	public IEditorInput findOriginalSourceForOuputFolderCopy(IEditorInput input) {
+		try {
+			IFile resource = ResourceUtil.getFile(input);
+			if (resource != null) {
+				// if the given resource is already pointing to a language file
+				if (fileExtensionProvider.isValid(resource.getFullPath().getFileExtension())) {
+					IJavaProject project = JavaCore.create(resource.getProject());
+					if (project.exists()) {
+						// Check if it's sitting in the output dir, look for original one
+						if (project.getOutputLocation().isPrefixOf(resource.getFullPath())) {
+							IPath relative = resource.getFullPath().removeFirstSegments(project.getOutputLocation().segmentCount());
+
+							List<IPackageFragmentRoot> packageFragmentRoots = new ArrayList<>();
+							for (IPackageFragmentRoot root : project.getPackageFragmentRoots()) {
+								if (root.getKind() == IPackageFragmentRoot.K_SOURCE) {
+									packageFragmentRoots.add(root);
+								}
+							}
+							
+							for (IPackageFragmentRoot source : packageFragmentRoots) {
+								IPath fullPath = source.getCorrespondingResource().getProjectRelativePath().append(relative);
+								IFile newFile = resource.getProject().getFile(fullPath);
+								if (newFile.exists()) {
+									return new FileEditorInput(newFile);
+								}
+							}
+						}
+
+						// Check if it's sitting in one of the output folders set on the source folders
+						List<IClasspathEntry> classpathEntries = new ArrayList<>();
```

Comment:
Populating a list to loop over the list on step later appears to be unnecessary. Maybe extract a few smaller methods to improve the readability and avoid the intermediate list here?